crates/iota-rest-kv/src/extractors.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module includes custom extractors needed for validation and custom
+//! errors messages on the input provided by the client
+
+use core::str;
+
+use axum::{
+    async_trait,
+    extract::{FromRequestParts, Path as AxumPath},
+    http::request::Parts,
+};
+use iota_storage::http_key_value_store::{Key, TaggedKey};
+use iota_types::{
+    digests::{CheckpointDigest, TransactionDigest},
+    storage::ObjectKey,
+};
+use serde::Deserialize;
+
+use crate::{errors::ApiError, types::ItemType};
+
+/// Path segment labels will be matched with struct field names
+#[derive(Deserialize, Debug)]
+struct DigestAndTypeParams {
+    /// The digest encoded as [`base64_url`]
+    digest: String,
+    /// The available supported items the digest can be associated with
+    item_type: ItemType,
+}
+
+/// We define our own extractor that includes validation and custom error
+/// message
+///
+/// This custom extractor matches Path segments and deserilize them internally
+/// into [`DigestAndTypeParams`] and constructs a [`Key`]
+pub struct ExtractPath(pub Key);
+
+#[async_trait]
+impl<S> FromRequestParts<S> for ExtractPath
+where
+    S: Send + Sync,
+{
+    type Rejection = ApiError;
+
+    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
+        match AxumPath::<DigestAndTypeParams>::from_request_parts(parts, state).await {
+            Ok(value) => {
+                // based on the item type construct the Key enum
+                let key = path_elements_to_key(&value.digest, value.item_type)?;
+
+                Ok(ExtractPath(key))
+            }
+            Err(e) => Err(ApiError::BadRequest(format!(
+                "invalid path parameter provided: {e}",
+            ))),
+        }
+    }
+}
+
+/// Create a a [`Key`] instance based on the provided [`base64_url`] encoded
+/// string and item type
+pub fn path_elements_to_key(digest: &str, item_type: ItemType) -> Result<Key, ApiError> {
+    let decoded_digest = base64_url::decode(digest)
```

Comment:
This can be also an object id or a sequence number, so I would suggest renaming to
```suggestion
    let decoded_key = base64_url::decode(digest)
```

Reply:
[fixed](https://github.com/iotaledger/iota/pull/5296/commits/689e31965cea777b8241d4621de5f8517c28f0d2#diff-69963076cf48ea415feec5a835104cb9658bd9ca7fc1e1df5994cd00cc09229dR64)